pkg/healthcheck/tcp.go

```diff
@@ -0,0 +1,142 @@
+package healthcheck
+
+import (
+	"context"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/traefik/traefik/v3/pkg/config/dynamic"
+	"github.com/traefik/traefik/v3/pkg/config/runtime"
+	"github.com/traefik/traefik/v3/pkg/tcp"
+)
+
+type ServiceTCPHealthChecker struct {
+	dialerManager *tcp.DialerManager
+	balancer      StatusSetter
+	info          *runtime.TCPServiceInfo
+
+	config   *dynamic.TCPServerHealthCheck
+	interval time.Duration
+	timeout  time.Duration
+
+	metrics metricsHealthCheck
+
+	targets     map[string]*net.TCPAddr
+	serviceName string
+}
+
+func NewServiceTCPHealthChecker(ctx context.Context, dialerManager *tcp.DialerManager, metrics metricsHealthCheck, config *dynamic.TCPServerHealthCheck, service StatusSetter, info *runtime.TCPServiceInfo, targets map[string]*net.TCPAddr, serviceName string) *ServiceTCPHealthChecker {
+	logger := log.Ctx(ctx)
+	interval := time.Duration(config.Interval)
+	if interval <= 0 {
+		logger.Error().Msg("Health check interval smaller than zero")
+		interval = time.Duration(dynamic.DefaultHealthCheckInterval)
+	}
+
+	timeout := time.Duration(config.Timeout)
+	if timeout <= 0 {
+		logger.Error().Msg("Health check timeout smaller than zero")
+		timeout = time.Duration(dynamic.DefaultHealthCheckTimeout)
+	}
+
```

Comment:
WDYT about limiting the payload length to prevent memory issues?
Same with validating that the expected response length is reasonable?